File: MapReduce_FastQ_BaseQual_Combiner/src/main/java/baseQual/FastQ_PerBaseQual_Reducer.java
package baseQual;

import org.apache.hadoop.io.DoubleWritable;
import org.apache.hadoop.io.IntWritable;
import utils.IdentifierPositionKeyWritable;
import utils.QualityCountHelperWritable;

import java.io.IOException;

/**
 * master-thesis Clemens Banas
 * Organization: DBIS - University of Innsbruck
 * Created 04.02.16.
 */
public class FastQ_PerBaseQual_Reducer extends org.apache.hadoop.mapreduce.Reducer<IdentifierPositionKeyWritable, QualityCountHelperWritable, IdentifierPositionKeyWritable, DoubleWritable> {

    @Override
    protected void reduce(IdentifierPositionKeyWritable key, Iterable<QualityCountHelperWritable> values, Context context) throws IOException, InterruptedException {
        context.write(key, getMeanValue(values));
    }

    private DoubleWritable getMeanValue(Iterable<QualityCountHelperWritable> values) {
        int sumOfQualityValues = 0;
        int numberOfQualityValues = 0;
        for (QualityCountHelperWritable qualCounter : values) {
            sumOfQualityValues += qualCounter.getSumOfQualityValues();
            numberOfQualityValues += qualCounter.getNumberOfQualityValues();
        }
        return new DoubleWritable((double)sumOfQualityValues / (double)numberOfQualityValues);
    }

}

File: Spark_FastQ_BaseQual_SingleFile_Baseline/src/main/scala/FastQ_PerBaseQual_Reducer.scala
package main.scala

import main.scala.utils.AvgCount

/**
  * master-thesis Clemens Banas
  * Organization: DBIS - University of Innsbruck
  * Created 21.04.2016
  */
object FastQ_PerBaseQual_Reducer {

  def createAverageCount(qualVal: Int): AvgCount = {
    new AvgCount(qualVal,1)
  }

  def addAndCount(a: AvgCount, qualVal: Int): AvgCount = {
    a.addQualVal(qualVal)
  }

  def combine(a: AvgCount, b: AvgCount): AvgCount = {
    new AvgCount(a.getSumOfQualityValues()+b.getSumOfQualityValues(), a.getNumberOfQualityValues()+b.getNumberOfQualityValues())
  }

}